modin/pandas/groupby.py

```diff
@@ -128,7 +128,10 @@ def sem(self, ddof=1):
         return self._default_to_pandas(lambda df: df.sem(ddof=ddof))
 
     def mean(self, *args, **kwargs):
-        return self._apply_agg_function(lambda df: df.mean(*args, **kwargs))
+        result = self._apply_agg_function(lambda df: df.mean(*args, **kwargs))
+        if self._by is None and not self._as_index:
+            result.reset_index(drop=True, inplace=True)
```

Comment:
It seems that this case is even more interesting. When grouping on an index level and performing reducing aggregation, the `as_index=False` parameter resets the index of the result to the default one without inserting anything to the actual frame:

<details><summary>Example</summary>

```python
>>> df
  by  col1  col2
0  a     1     3
1  a     2     4
2  b     3     5
3  b     4     6
>>> df.set_index("by").groupby(level=0, as_index=False).sum()
   col1  col2
0     3     7
1     7    11
>>> df.set_index("by").groupby(level=0, as_index=False).nunique()
   col1  col2
0     2     2
1     2     2
>>> df.set_index("by").groupby(level=0, as_index=False).apply(lambda df: df.sum())
   col1  col2
0     3     7
1     7    11
```

</details>

But when applying a map-aggregation, the `as_index=False` parameter has just no effect, because it always trying to preserve the index of the original frame:
<details><summary>Example</summary>

```python
>>> df.set_index("by").groupby(level=0, as_index=False).apply(lambda df: df + 1)
    col1  col2
by
a      2     4
a      3     5
b      4     6
b      5     7
>>> df.set_index("by").groupby(level=0, as_index=False).shift()
    col1  col2
by
a    NaN   NaN
a    1.0   3.0
b    NaN   NaN
b    3.0   5.0
>>> df.set_index("by").groupby(level=0, as_index=False).fillna(0)
    col1  col2
by
a      1     3
a      2     4
b      3     5
b      4     6
# Grouping only on a single index level, but both are presented in the result
>>> df.set_index(["by", "col1"]).groupby(level=0, as_index=False).fillna(0)
         col2
by col1
a  1        3
   2        4
b  3        5
   4        6
```

</details>

A proper solution would be is to follow the pandas logic: always reset index when grouping on index level and restore it only for a certain non-reducing aggregations at the API layer. But this approach may slow-down groupby for map-aggregations since frame relabeling might be quite expensive (especially if the frame has a lot of delayed computations). 

Another solution could be is to deduce somehow whether the aggregation is reducing or not at the place where we're dealing with `as_index` (at the QC), for example, we can check, whether the length of the result has changed after aggregation was applied (but it looks like an easy-breakable solution).

I'd like to keep the logic of handling `as_index` in one place, not to split it between backend and front-end (ideally, we want to see the whole `as_index` logic at the front, since this parameter is a tricky one and we don't want to delegate such pandas-dependent task to the backend), that's why I'm complaining about the current solution. But let's wait for other reviewers' opinions, maybe it's an acceptable compromise.

BTW, the first steps of moving the handling of `as_index` to the front were made in #3592, where I encapsulate the trickiest of `as_index` processing in a single backend-agnostic function, that actually could be used at the front-end. You can take a look at it, maybe you would have some suggestions of how to modify it in order to support these grouping-on-index-level cases.

Reply:
Looks like refactoring related to moving of `as_index` processing to the front-end is a separate issue, which should be implemented after this PR and #3592. What do you think?

Reply:
Did you create an issue?

Reply:
A comment should probably be added to the code on why we reset index. That will help to handle/understand the code for such corner cases in future.

Reply:
Issue related to moving the handling of `as_index` is here #3716